srf_generation/Fault.py

```diff
@@ -0,0 +1,492 @@
+        located half way along strike and downdip
+        2) The fault area is computed from the median of a Moment scaling
+        relationship and thus is assumed to be a 'typical' stress drop event for
+        the Mw-relationship used.
+        Should only be used for small events where the downdip width is smaller
+        than the seismogenic depth (i.e. so the assumption of a square fault plane
+        is reasonable); depth is reset to zero if above ground values determined
+        3) srike in deg measured clockwise from N; rake in deg measured
+        anti-clockwise from the strike direction (and in the range
+        -180<lambda<180; dip in deg measured downward from horizontal
+
+        Input variables:
+        Lat    - the latitude of the focal mechanism (-ve below equator)
+        Lon    - the lon of the focal mech (-180<lon<180)
+        Depth  - the depth of the focal mech (in km)
+        Mw     - the moment magnitude from the Mw tensor soln
+        strike - the strike of the focal mech (only 1)
+        rake   - rake (not used, but carried forward)
+        dip    - the dip angle of the fault plane
+
+        output variables:
+        lat       - the latitude of the subfault
+        lon       - the lon of the subfault
+        depth     - depth of the subfault
+        lonAsList - as a 1D array
+        """
+        if self.magnitude_scaling_relation is None:
+            raise ValueError("No magnitude scaling relation given")
+        # rounded subfault spacing
+
+        # use cartesian coordinate system to define the along strike and downdip
+        # locations taking the center of the fault plane as (x,y)=(0,0)
+        y_pos: np.ndarray = np.arange(self.dwid / 2.0, self.width, self.dwid)[::-1] - self.width / 2.0
+
+        depth_loc_relative = (-y_pos * np.sin(np.radians(self._dip))).repeat(self.nx).reshape((self.nx, self.ny))
+        shift = np.min(self.dtop + depth_loc_relative, 0)
+        self.depths = self.dtop + depth_loc_relative - shift
+        if np.any(0 > self.depths):
+            raise ValueError(
+                "Some points are above the ground. This represents a logic problem and should be referred to the "
+                "developers."
+            )
+
+
+class Type3(FiniteFault):
+    def __init__(self, name, magnitude, lon1, lat1, lon2, lat2, dip, rake, dtop, dbottom, dip_dir):
+        self._trace = ((lon1, lat1), (lon2, lat2))
+        self._dtop = dtop
+        self._dbottom = dbottom
+        self._dip_dir = dip_dir
+
+        self._length = geo.ll_dist(lon1, lat1, lon2, lat2)
+        self._width = (dbottom - dtop)/np.sin(dip)
+
+        strike = geo.ll_bearing(lon1, lat1, lon2, lat2)
+
+        super().__init__(name, magnitude, strike, rake, dip)
+
+    def to_dict(self):
+        base_dict: dict = super().to_dict()
+        base_dict.update({
+            "type": 3,
+        })
+        return base_dict
+
+
+class Type4(MultiPlaneFault):
```

Comment:
I think each plane needs an init time, but that can be an extension later.

Reply:
Yeah, Type3/4 are just an outline for when NHM faults get implemented